src/ValidatingHydrator.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Input\Validation;
+
+use Yiisoft\Hydrator\HydratorInterface;
+use Yiisoft\Input\Validation\Attribute\EarlyValidationResolver;
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\ValidatorInterface;
+
+final class ValidatingHydrator implements HydratorInterface
+{
+    public function __construct(
+        private readonly HydratorInterface $hydrator,
+        private readonly ValidatorInterface $validator,
+        private readonly EarlyValidationResolver $earlyValidationResolver,
+    ) {
+    }
+
+    public function hydrate(
+        object $object,
+        array $data = [],
+        array $map = [],
+        bool $strict = false
+    ): void {
+        $result = $this->beforeAction();
+        $this->hydrator->hydrate($object, $data, $map, $strict);
+        $this->afterAction($object, $result);
+    }
+
+    public function create(
+        string $class,
+        array $data = [],
+        array $map = [],
+        bool $strict = false
+    ): object {
+        $result = $this->beforeAction();
+        $object = $this->hydrator->create($class, $data, $map, $strict);
```

Comment:
The first line is a wrap for specific attribute handler, so every same packages need to write the same wrappers to get validation result. 
As far as I see, EaryValidation goes through the hydrator pipeline. What if hydrator expose validation/hydration results by flag / separate method / other?

`[$object, $hydrationResult] = $hydrator->create(...)` – two result values
`[$object, $hydrationResult] = $hydrator->createAndGetResult(...)` – the same, but with an additional method
`$object = $hydrator->create(..., &$hydrationResult)` – bind a new variable as `preg_match` do
`$object = $hydrator->create(..., strict: true)` – the result errors should be always empty 
`$object = $hydrator->create(...); $result = $resultProvider->getResult($object)` – original flow with stateful result provider

Reply:
Has use cases for this suggestion?

Reply:
Decided not to do this.